datafusion/sql/src/unparser/dialect.rs

```diff
@@ -15,19 +15,33 @@
 // specific language governing permissions and limitations
 // under the License.
 
-/// Dialect is used to capture dialect specific syntax.
+use regex::Regex;
+use sqlparser::keywords::ALL_KEYWORDS;
+
+/// `Dialect` to use for Unparsing
+///
+/// The default dialect tries to avoid quoting identifiers unless necessary (e.g. `a` instead of `"a"`)
+/// but this behavior can be overridden as needed
 /// Note: this trait will eventually be replaced by the Dialect in the SQLparser package
 ///
 /// See <https://github.com/sqlparser-rs/sqlparser-rs/pull/1170>
 pub trait Dialect {
     fn identifier_quote_style(&self) -> Option<char>;
+    fn identifier_needs_quote(&self, _: &str) -> bool {
```

Comment:
Above note said this trait will eventually be replaced by the Dialect in the SQLparser package. Seems this pr make this harder. Should we extend sqlparser `Dialect` using something like `DialectExt` trait?

Reply:
I wanted to note that this functionality could also be covered within the existing `SQLparser::Dialect::identifier_quote_style`. It's signature looks as follows:
```
identifier_quote_style(&self, _identifier: &str) -> Option<char>
```
It is passed the `identifier` and can optionally return a quote character if needed. This way the trait doesn't need extending at all.

Reply:
@goldmedal  let me know what you want to do here -- I can merge this PR and we can update this per @backkem 's suggestion in a follow on PR, or would you like to update this PR?

Reply:
> @goldmedal let me know what you want to do here -- I can merge this PR and we can update this per @backkem 's suggestion in a follow on PR, or would you like to update this PR?
Thanks @lewiszlw @backkem @alamb 
I think I have time to fix it now. I can fix it in this PR.